File: app/src/main/java/com/example/suraksha_ai/services/AIDangerService.kt
package com.example.suraksha_ai.services

import android.graphics.Bitmap
import kotlinx.coroutines.delay
import kotlin.random.Random
import kotlin.math.abs

class AIDangerService {
    private var frameCount = 0
    private var lastDangerScore = 0.0
    private val dangerHistory = mutableListOf<Double>()
    
    // Simulate realistic danger patterns for demo
    suspend fun analyzeFrame(bitmap: Bitmap?): Double {
        delay(80) // Simulate processing time
        
        frameCount++
        
        // Create more realistic danger patterns
        val random = Random(System.currentTimeMillis())
        
        // Base danger level (usually low)
        var dangerScore = random.nextDouble() * 0.25
        
        // Simulate occasional spikes (like someone approaching)
        if (frameCount % 30 == 0 && random.nextDouble() < 0.15) {
            // Simulate threat approaching
            dangerScore = 0.5 + (random.nextDouble() * 0.3)
        }
        
        // Simulate sustained threat (following behavior)
        if (dangerHistory.isNotEmpty() && dangerHistory.last() > 0.5) {
            dangerScore = (dangerHistory.last() * 0.8) + (random.nextDouble() * 0.2)
        }
        
        // Add some continuity (danger doesn't jump randomly)
        if (dangerHistory.isNotEmpty()) {
            val lastScore = dangerHistory.last()
            dangerScore = (lastScore * 0.7) + (dangerScore * 0.3)
        }
        
        // Keep history (last 10 frames)
        dangerHistory.add(dangerScore)
        if (dangerHistory.size > 10) {
            dangerHistory.removeAt(0)
        }
        
        lastDangerScore = dangerScore
        return dangerScore.coerceIn(0.0, 1.0)
    }

    fun getDangerDetails(dangerScore: Double): DangerDetails {
        return when {
            dangerScore < 0.3 -> DangerDetails(
                level = "Safe",
                color = 0xFF4CAF50,
                threats = emptyList()
            )
            dangerScore < 0.6 -> DangerDetails(
                level = "Caution",
                color = 0xFFFF9800,
                threats = listOf("Person detected nearby")
            )
            else -> DangerDetails(
                level = "Danger",
                color = 0xFFDC143C,
                threats = listOf(
                    "Stranger too close",
                    "Possible following detected",
                    "Aggressive behavior detected"
                )
            )
        }
    }
}

data class DangerDetails(
    val level: String,
    val color: Long,
    val threats: List<String>
)

File: lib/services/ai_danger_service.dart
import 'dart:math';

class AIDangerService {
  // Placeholder for ML model inference
  // In production, this would use TensorFlow Lite or similar
  
  Future<double> analyzeFrame({
    required List<int> imageData,
    required int width,
    required int height,
  }) async {
    // Simulate AI analysis with random danger score
    // Real implementation would:
    // 1. Preprocess image
    // 2. Run through ML model
    // 3. Detect: strangers, following, expressions, weapons, groups
    // 4. Return danger score 0.0 - 1.0
    
    await Future.delayed(const Duration(milliseconds: 100));
    
    // Simulated danger detection
    final random = Random();
    final baseDanger = random.nextDouble() * 0.3; // Base 0-0.3
    
    // Simulate occasional high danger
    if (random.nextDouble() < 0.05) {
      return 0.7 + (random.nextDouble() * 0.3); // 0.7-1.0
    }
    
    return baseDanger;
  }

  Map<String, dynamic> getDangerDetails(double dangerScore) {
    if (dangerScore < 0.3) {
      return {
        'level': 'Safe',
        'color': 0xFF4CAF50,
        'threats': [],
      };
    } else if (dangerScore < 0.6) {
      return {
        'level': 'Caution',
        'color': 0xFFFF9800,
        'threats': ['Person detected nearby'],
      };
    } else {
      return {
        'level': 'Danger',
        'color': 0xFFDC143C,
        'threats': [
          'Stranger too close',
          'Possible following detected',
          'Aggressive behavior detected',
        ],
      };
    }
  }
}

